enigma/src/main/java/org/quiltmc/enigma/api/translation/mapping/serde/MappingFormat.java

```diff
@@ -90,34 +96,47 @@ public MappingsReader getReader() {
 		return this.reader;
 	}
 
+	public FileType getFileType() {
+		return this.fileType;
+	}
+
 	/**
-	 * Determines the mapping format of the provided file. Checks all formats, and returns {@link #PROGUARD} if none match.
+	 * Determines the mapping format of the provided file. Checks all formats according to their {@link #getFileType()} file extensions.
+	 * If the file is a directory, it will check the first file in the directory.
+	 * Will return {@link #PROGUARD} if no format is found for single files, and {@link #ENIGMA_DIRECTORY} if no format is found for directories.
 	 * @param file the file to analyse
-	 * @apiNote Any directory is considered to be the {@link #ENIGMA_DIRECTORY} format.
-	 * Proguard does not have an explicit file extension, so it is the fallback.
 	 * @return the mapping format of the file.
 	 */
 	public static MappingFormat parseFromFile(Path file) {
 		if (Files.isDirectory(file)) {
-			return ENIGMA_DIRECTORY;
-		} else {
-			switch (MoreFiles.getFileExtension(file).toLowerCase()) {
-				case "zip" -> {
-					return ENIGMA_ZIP;
-				}
-				case "mapping", "mappings" -> {
-					return ENIGMA_FILE;
-				}
-				case "tiny" -> {
-					return TINY_V2;
-				}
-				case "tsrg" -> {
-					return SRG_FILE;
+			try {
+				File firstFile = Arrays.stream(Objects.requireNonNull(file.toFile().listFiles())).findFirst().orElseThrow();
+
+				for (MappingFormat format : values()) {
+					if (!format.getFileType().isDirectory()) {
+						continue;
+					}
+
+					String extension = MoreFiles.getFileExtension(firstFile.toPath()).toLowerCase();
+					if (format.fileType.getExtensions().contains(extension)) {
+						return format;
+					}
 				}
-				default -> {
-					return PROGUARD;
+
+				return ENIGMA_DIRECTORY;
+			} catch (Exception e) {
+				return ENIGMA_DIRECTORY;
+			}
+		} else {
+			String extension = MoreFiles.getFileExtension(file).toLowerCase();
+
+			for (MappingFormat format : values()) {
+				if (format.fileType.getExtensions().contains(extension)) {
+					return format;
 				}
 			}
 		}
+
+		return PROGUARD;
```

Comment:
hm i dont like this default. seem like it could break stuff

Reply:
proguard doesn't have an explicitly defined file extension -- this matches old behaviour